api/src/main/java/org/openmrs/module/commonreports/reports/VaccinationReportManager.java

```diff
@@ -0,0 +1,209 @@
+		params.add(getEndDateParameter());
+		return params;
+	}
+	
+	@Override
+	public ReportDefinition constructReportDefinition() {
+		
+		ReportDefinition rd = new ReportDefinition();
+		rd.setUuid(getUuid());
+		rd.setName(getName());
+		rd.setDescription(getDescription());
+		
+		rd.setParameters(getParameters());
+		
+		CohortCrossTabDataSetDefinition vaccination = new CohortCrossTabDataSetDefinition();
+		vaccination.addParameters(getParameters());
+		rd.addDataSetDefinition(getName(), Mapped.mapStraightThrough(vaccination));
+		
+		Map<String, Object> parameterMappings = new HashMap<String, Object>();
+		parameterMappings.put("onOrAfter", "${startDate}");
+		parameterMappings.put("onOrBefore", "${endDate}");
+		
+		String[] vaccinationConceptsListWithSequence = inizService
+		        .getValueFromKey("report.vaccination.vaccinationConceptsListWithSequence").split(",");
+		
+		for (String member : vaccinationConceptsListWithSequence) {
+			
+			String[] bits = member.split(":");
+			String lastOne = bits[bits.length - 1];
+			if (!NumberUtils.isNumber(lastOne)) {
+				String sqlQuery = "SELECT person_id FROM obs where obs_group_id IN (SELECT obs_group_id FROM obs where concept_id= (select DISTINCT CONCEPT_ID from concept where uuid='"
+				        + inizService.getValueFromKey("report.vaccination.vaccinations")
+				        + "') and value_coded=(select DISTINCT CONCEPT_ID from concept where uuid='" + member + "'));";
+				SqlCohortDefinition sql = new SqlCohortDefinition(sqlQuery);
+				sql.addParameter(new Parameter("onOrAfter", "On Or After", Date.class));
+				sql.addParameter(new Parameter("onOrBefore", "On Or Before", Date.class));
+				
+				vaccination.addRow(conceptService.getConceptByUuid(member).getDisplayString(), sql, parameterMappings);
+				
+			} else {
+				int lastIndex = Integer.parseInt(lastOne);
+				String vacName = member.substring(0, member.lastIndexOf(":"));
+				
+				String sqlQuery = "SELECT person_id FROM obs where obs_group_id IN (SELECT obs_group_id FROM obs where concept_id= (select DISTINCT CONCEPT_ID from concept where uuid='"
+				        + inizService.getValueFromKey("report.vaccination.vaccinations")
+				        + "') and value_coded=(select DISTINCT CONCEPT_ID from concept where uuid='" + vacName
+				        + "')) AND ((concept_id=(select DISTINCT CONCEPT_ID from concept where uuid='"
+				        + inizService.getValueFromKey("report.vaccination.vaccinationSequenceNumberConcept")
+				        + "') AND value_numeric=" + lastIndex
+				        + ") OR (concept_id=(select DISTINCT CONCEPT_ID from concept where uuid='"
+				        + inizService.getValueFromKey("report.vaccination.boostervaccinationSequenceNumberConcept")
+				        + "') AND value_numeric=" + lastIndex + "))";
+				SqlCohortDefinition sql = new SqlCohortDefinition(sqlQuery);
+				sql.addParameter(new Parameter("onOrAfter", "On Or After", Date.class));
+				sql.addParameter(new Parameter("onOrBefore", "On Or Before", Date.class));
```

Comment:
I'd use a `CodedObsCohortDefinition` or `NumericObsCohortDefinition` here to minimise sql queries as much as possible. Related to https://github.com/mekomsolutions/openmrs-module-commonreports/pull/30#discussion_r674812833